### PR TITLE
fix(gatsby): context.nodeModel.runQuery should return [] instead of null

### LINKED ISSUE #25857: context.nodeModel.runQuery should return [] instead of null
## Description

With a query that returns no results, `context.nodeModel.runQuery` returns `Promise<null>`, but [the documentation](https://www.gatsbyjs.org/docs/node-model/#runQuery) specifies non-null type `Promise<Node[]>`.

I'd prefer to get back `[]` instead of `null`, because dealing with one type is easier than dealing with two, and because it's consistent with how the rest of Gatsby GraphQL works. If I pass the same filter to a static query and `runQuery`, I think both should return the same result, `[]`.

### Environment

```txt
  System:
    OS: Linux 4.19 Debian GNU/Linux 10 (buster) 10 (buster)
    CPU: (12) x64 Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz
    Shell: 5.0.3 - /bin/bash
  Binaries:
    Node: 12.16.3 - /usr/local/bin/node
    Yarn: 1.22.4 - /usr/bin/yarn
    npm: 6.14.4 - /usr/local/bin/npm
  Languages:
    Python: 2.7.16 - /usr/bin/python
  npmPackages:
    gatsby: ^2.24.4 => 2.24.4
    gatsby-image: ^2.4.13 => 2.4.13
    gatsby-plugin-google-gtag: ^2.1.10 => 2.1.10
    gatsby-plugin-manifest: ^2.4.18 => 2.4.18
    gatsby-plugin-netlify: ^2.3.11 => 2.3.11
    gatsby-plugin-react-helmet: ^3.3.10 => 3.3.10
    gatsby-plugin-sharp: ^2.6.19 => 2.6.19
    gatsby-plugin-theme-ui: ^0.3.0 => 0.3.0
    gatsby-source-sanity: ^6.0.3 => 6.0.3
    gatsby-theme-style-guide: ^0.3.1 => 0.3.1
    gatsby-transformer-sharp: ^2.5.11 => 2.5.11
  npmGlobalPackages:
    gatsby-cli: 2.12.45
```

### PR DESCRIPTION
**Pull request by kartikcho**
As of now, context.nodeModel.runQuery returns `Promise<null>`, but the [documentation](https://www.gatsbyjs.org/docs/node-model/#runQuery) specifies type `Promise<Node[]>` to be returned. 

This PR resolves #25857 by updating the default return statement.

### CODE REVIEW

**Comment by kartikcho** [comment]
@vladar could you provide more context on these tests that seem to affect the change in this PR. Thanks!

https://github.com/gatsbyjs/gatsby/blob/09b36886d5a392ea15ce3f5ee29b49f2b9599010/packages/gatsby/src/schema/__tests__/run-query.js#L747-L749

https://github.com/gatsbyjs/gatsby/blob/09b36886d5a392ea15ce3f5ee29b49f2b9599010/packages/gatsby/src/schema/__tests__/run-query.js#L942-L944

**Comment by vladar** [comment]
Thanks for the PR! I think you should modify those tests to expect `[]` vs. `null` (as that's effectively what we change here).

But as I mentioned in the issue this is likely a breaking change, so we'll keep this PR open until the next major.

File: packages/gatsby/src/redux/run-fast-filters.ts [hunk 1]
```diff
@@ -420,7 +420,7 @@ function convertAndApplyFastFilters(
   if (firstOnly) {
     return []
   }
```
Comment (vladar) [inline packages/gatsby/src/redux/run-fast-filters.ts:422]:
```suggestion
```

File: packages/gatsby/src/redux/run-fast-filters.ts [hunk 2]
```diff
@@ -417,10 +417,7 @@ function convertAndApplyFastFilters(
     stats.totalSiftHits++
   }
 
-  if (firstOnly) {
```
Comment (pvdz) [inline packages/gatsby/src/redux/run-fast-filters.ts:420]: iirc there were one or two other places where this happens (returning `null` vs returning an empty array) but maybe I already unified those into this check here.

**Comment by pvdz** [comment]
I'm going to put this PR in draft mode as we won't be merging it any time soon. Rest assured that I really want this PR merged once we can!

**Comment by muescha** [comment]
should then the docs be updated to reflect the current state (with a deprecation warning that it would change in 3.0)?

**Comment by pvdz** [comment]
It's been like this for so long that I would kind of suggest to leave it as is for now. But if you're set on updating them then you could add a note that we plan to eliminate this inconsistency in a future major bump.

File: packages/gatsby/src/redux/run-fast-filters.ts (hunk 2 above)
Comment (kartikcho) [reply packages/gatsby/src/redux/run-fast-filters.ts:420]: What do you want to update here @pvdz?

**Comment by kartikcho** [comment]
Thanks for still working on getting this in @vladar, let me know if I can be of help in any way!

**Review by vladar — APPROVED** [review]
Thanks for the PR and let's merge it, finally! 🎉

**Pull request merged**